llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp

```diff
@@ -1960,7 +1960,9 @@ void AsmPrinter::emitFunctionBody() {
   // are automatically sized.
   bool EmitFunctionSize = MAI->hasDotTypeDotSizeDirective() && !TT.isWasm();
 
-  if (EmitFunctionSize || needFuncLabels(*MF, *this)) {
+  // SPIR-V supports label instructions only inside a block, not after the
+  // function body.
+  if (!TT.isSPIRV() && (EmitFunctionSize || needFuncLabels(*MF, *this))) {
```

Comment:
Can't emit triple checks like this in generic code. Isn't there some override in needFuncLabels?

Reply:
Thank you @arsenm
Please see two lines above (line 1961) in this very general code a check for TT.isWasm(), showing similar approach as proposed in the PR.

As for needFuncLabels(), no, it has no overrides. It always returns true if valid debug info is present, and valid debug info is encoded by a private class member, thus forbidding any workarounds and on/off manipulations over valid debug info status (that actually also would be a worse hack that proposed in the PR).

Please see my comment to https://github.com/llvm/llvm-project/issues/102732 for more words about the problem, but the problem is pretty simple, there are no existing hooks/overrides that allow to keep SPIR-V valid due to this check for debug info in AsmPrinter.

Reply:
Let's recap it:
1) AsmPrinter generates invalid SPIR-V because needFuncLabels() is inside the OR condition and returns true, and needFuncLabels() is a static function,
2) needFuncLabels() returns true because Asm.hasDebugInfo() is inside the OR condition and returns true, and Asm.hasDebugInfo() just returns DbgInfoAvailable,
3) DbgInfoAvailable is a private class member and it's true if valid debug info is present.
4) In the very same code (actually, for the very same if) we have a check for the triple in `TT.isWasm()`.

Reply:
Is this really a property of the target, or the SPIRV object format? It would be less objectionable if it was an object format check, and not the target

Reply:
I'd say that in case of SPIR-V we have a peculiar situation, when we should not talk about the target per se, separately from the format. The reason for this is that SPIR-V is a higher level representation, as LLVM IR, and this differs it from other lower-level backends.

In case of SPIR-V both the SPIR-V target (behavior/properties) and the SPIR-V representation (format) are artefacts of the SPIR-V specification -- the SPIR-V target does not have in mind any *real* target, it's just an intermediate format, as abstract as the format is.

The SPIR-V specification dictates that a label may appear only inside the block (and so not outside of a function). Given the abstract nature and full congruence of SPIR-V target and format, there is no and will never be a format for the SPIR-V target that allows labels outside of a function, because the format in this case is a material embodiment of the abstract target.

Reply:
Does this my rationale makes sense for you @arsenm ? Thank you

Reply:
@arsenm Gentle ping

Reply:
@arsenm I understand and respect the intent to preserve AsmPrinter agnostic toward targets. In this case, however, AsmPrinter doesn't conform to specification requirements of the SPIR-V target.

Having AsmPrinter and MCStreamer with the hardcoded emit-label functionality without any way to override the default behavior for Modules with debug info, having MCAsmStreamer marked as `final` and having no way to register an alternative MCStreamer, this PR seems an adequate way to resolve the issue and doesn't seem controversial.

I don't think that this issue is worth adding a hook to AsmPrinter.

Reply:
The target and object format may be intricately linked, but we still have a defined SPIRV ObjectFormatType. So you can just change the getArch() == spirv check for a getObjectFormat() == spirv check?

Reply:
Yes, indeed. Thank you for the suggestion. The check is updated to test the SPIRV object format.